### ISSUE TITLE
Respect return value of 'pre' as new payload

### ISSUE THREAD

**Issue by tripodsan** [adobe/parcel-plugin-htl#23]
In order to avoid parameter re-assign, it should be possible to return the new payload. eg:

```
function pre(payload) {
  const myPayload = Object.assign({}, payload);
  myPayload.resource.time = `${new Date()}`;
  return myPayload;
}
```

**Comment by trieloff** [comment]
In case you do both: parameter re-assign and `return`, the return values takes precedence.

**Comment by tripodsan** [comment]
> In case you do both: parameter re-assign and return, the return values takes precedence.

yes. I think it would be very hard to implement it otherwise :-)

**Comment by kptdobe** [comment]
The `no-param-reassign` is rule is annoying and only useful to avoid errors if you use the `arguments` array. See https://spin.atomicobject.com/2011/04/10/javascript-don-t-reassign-your-function-arguments/

In our simple case, using a new variable is enough, no need to copy the whole object (this is a recommended option in https://eslint.org/docs/rules/no-param-reassign): 

`const myPayload = payload;` is way enough!!

**Comment by tripodsan** [comment]
> const myPayload = payload; is way enough!!

well, it does change the input object. If you want to be save not to change the object you need to deeply clone it. however, I think our pipeline approach encourages to change the "payload" as it traverses the pipeline.

**Comment by kptdobe** [comment]
Yes, but that's not why the rule has been made and does not protect you to change the input object anyway.
Because of that rule, we introduce the idea that the payload object should not be changed while this is not necessarly in the "API" definition... If it becomes, we then need to test accordingly.
And it makes the code more complex...

We should get rid of that rule to keep the code simple!! ;)

**Comment by tripodsan** [comment]
> We should get rid of that rule to keep the code simple!! ;)

exactly. we don't include this rule anyways, as we don't enforce any linting for developer projects.

**Comment by kptdobe** [comment]
Agreed. 
Pb is the sample code you wrote and that will be used as a basis for developers that will copy/paste it and rely on the "clone". ;)
